lib/ukstore/store.c

```diff
@@ -29,3 +29,1194 @@
+	GETCASE_UPCAST(e, s32, S32, out, s32);
+	GETCASE_DOWNCASTUS(e, u64, S32, out, s32);
+	GETCASE_DOWNCASTSS(e, s64, S32, out, s32);
+	GETCASE_DOWNCASTUS(e, uptr, S32, out, s32);
+
+	case uk_store_get_entry_type(charp): {
+		char *val = NULL;
+		__s64 nr;
+
+		ret = e->get.charp(&val);
+		if (ret < 0)
+			return ret;
+		nr = atoi(val);
+		if (unlikely(nr < __S32_MIN || nr > __S32_MAX))
+			return -ERANGE;
+		*out  = (__s32) nr;
+		return ret;
+	}
+
+	default:
+		return -EINVAL;
+	}
+}
+
+int
+_uk_store_get_u64(struct uk_store_entry *e, __u64 *out)
+{
+	int ret;
+
+	UK_ASSERT(e);
+
+	switch (e->type) {
+	GETCASE_UPCAST(e, u8, U64, out, u64);
+	GETCASE_UPCASTSU(e, s8, U64, out, u64);
+	GETCASE_UPCAST(e, u16, U64, out, u64);
+	GETCASE_UPCASTSU(e, s16, U64, out, u64);
+	GETCASE_UPCAST(e, u32, U64, out, u64);
+	GETCASE_UPCASTSU(e, s32, U64, out, u64);
+	GETCASE_UPCAST(e, u64, U64, out, u64);
+	GETCASE_UPCASTSU(e, s64, U64, out, u64);
+	GETCASE_UPCAST(e, uptr, U64, out, u64);
+
+	case uk_store_get_entry_type(charp): {
+		char *val = NULL;
+		__s64 nr;
+
+		ret = e->get.charp(&val);
+		if (ret < 0)
+			return ret;
+		nr = atoi(val);
+		if (unlikely(nr < 0))
+			return -ERANGE;
+		*out  = (__u64) nr;
+		return ret;
+	}
+
+	default:
+		return -EINVAL;
+	}
+}
+
+int
+_uk_store_get_s64(struct uk_store_entry *e, __s64 *out)
+{
+	int ret;
+
+	UK_ASSERT(e);
+
+	switch (e->type) {
+	GETCASE_UPCAST(e, u8, S64, out, s64);
+	GETCASE_UPCAST(e, s8, S64, out, s64);
+	GETCASE_UPCAST(e, u16, S64, out, s64);
+	GETCASE_UPCAST(e, s16, S64, out, s64);
+	GETCASE_UPCAST(e, u32, S64, out, s64);
+	GETCASE_UPCAST(e, s32, S64, out, s64);
+	GETCASE_DOWNCASTUS(e, u64, S64, out, s64);
+	GETCASE_UPCAST(e, s64, S64, out, s64);
+	GETCASE_DOWNCASTUS(e, uptr, S64, out, s64);
+
+	case uk_store_get_entry_type(charp): {
+		char *val = NULL;
+		__s64 nr;
+
+		ret = e->get.charp(&val);
+		if (ret < 0)
+			return ret;
+		nr = atoi(val);
+		*out  = (__s64) nr;
+		return ret;
+	}
+
+	default:
+		return -EINVAL;
+	}
+}
+
+int
+_uk_store_get_uptr(struct uk_store_entry *e, __uptr *out)
+{
+	int ret;
+
+	UK_ASSERT(e);
+
+	switch (e->type) {
+	GETCASE_UPCAST(e, u8, PTR, out, uptr);
+	GETCASE_UPCASTSU(e, s8, PTR, out, uptr);
+	GETCASE_UPCAST(e, u16, PTR, out, uptr);
+	GETCASE_UPCASTSU(e, s16, PTR, out, uptr);
+	GETCASE_UPCAST(e, u32, PTR, out, uptr);
+	GETCASE_UPCASTSU(e, s32, PTR, out, uptr);
+	GETCASE_UPCAST(e, u64, PTR, out, uptr);
+	GETCASE_UPCASTSU(e, s64, PTR, out, uptr);
+	GETCASE_UPCAST(e, uptr, PTR, out, uptr);
+
+	case uk_store_get_entry_type(charp): {
+		char *val = NULL;
+		__s64 nr;
+
+		ret = e->get.charp(&val);
+		if (ret < 0)
+			return ret;
+		nr = atoi(val);
+		if (unlikely(nr < 0))
+			return -ERANGE;
+		*out  = (__uptr) nr;
+		return ret;
+	}
+
+	default:
+		return -EINVAL;
+	}
+}
+
+int
+_uk_store_get_charp(struct uk_store_entry *e, char **out)
+{
+	int ret;
+	char *str;
+
+	UK_ASSERT(e);
+
+	str = calloc(UK_STORE_NR_MAX_LEN, sizeof(char));
```

Comment:
You probably should allocate within the cases and have as many digits a number can maximum have including the sign. If we enable user provided buffers with size, we can use `snprintf`